YACS-RCOS/yacs.n
# Add hidden semesters & remove all data then reinsert on CSV upload

## THIS PR IS READY FOR REVIEW

Code is a little messy atm, focusing on the feature working rn.

Fixed user account logout, added user tables & semester_info to build script. Non-public semesters are hidden to all users except admins in the footer UI.

- [x] Need to wipe courses & course_session on CSV upload prior to insertion
- [X] Need to protect routes for semesters that are specified to be non-public.
![image](https://user-images.githubusercontent.com/28029704/79680969-88266b80-81e3-11ea-9be4-6dbc9ea1ff97.png)

- [X] Fix client-side filtering of all db courses by parameterizing query by semester
![image](https://user-images.githubusercontent.com/28029704/79680937-2fef6980-81e3-11ea-9a19-ee2274631bca.png)
- [X] Should reconsider what the 'default' semester is. Choose the first public one?

- [X] Changed footer anchor tags to btns that will update the current page. This means an API request for each time you change semesters and you will see some delay. Am not sure how big of a deal this will be. To make this prettier, I believe there is a way to 'fake' changing the page in a SPA app (so the url changes) but the state of the SPA changes. Wasn't too concerned with look for this PR however.
![image](https://user-images.githubusercontent.com/28029704/79702087-19dbba80-8270-11ea-82b8-06b6f36b817b.png)



**NOTE**

I am using [flask sessions](https://flask.palletsprojects.com/en/1.1.x/quickstart/?highlight=middleware#sessions) instead of the `user_session` tbl. Flask session makes it easier for me to persist the logged in user data rather than having to query the session tbl for a uid then query users with that uid. If this is a problem, it shouldn't be too hard to refactor.

In this PR, the sessionID that is part of user_sessions table serves only the purpose of recording log ins and log outs.

**How to test**:

1. Rebuild DB via `src/db/build.sh`
2. Create an admin user via psql
```
INSERT INTO user_account VALUES (
    1,
    'admin',
    'admin@rpi.edu',
    '2904092',
    'f52fbd32b2b3b86ff88ef6c490628285f482af15ddcb29541f94bcf526a3f6c7',
    'cs',
    'under',
    1::boolean,
    1::boolean,
    1::boolean
);
```
3. Go to `/admin` and upload two CSVs where one of them is non-public, can do this at upload stage or in `Edit Semesters` action on admin panel.
4. Go to schedule page and look at footer and make sure the non-public link isn't there.
5. Go to the non-public link via URL and you should get an HTTP 401 unauthorized error.
6. Login with user "admin" and password "hunter2"
7. The non-public link in the footer should be there.
8. The course data should be loaded for that semester.
9. Press logout and a timestamp should be added to `user_session` tbl and your session cookies will be cleared.

**IMPROVEMENTS THAT COULD BE MADE**
- Separate page for unauthorized request? Or some indication beyond dev tools?
- Indication that currently looking at a non-public semester on the schedule UI.
- Make Edit Semesters component more robust; separate out validation logic, tweak organization.

Closes #105
Closes #110
